src/pyrokinetics/gk_code/gkw.py

```diff
@@ -0,0 +1,1331 @@
+        returns a dictionary of the different reference species. If the
+        references used match the default references then an empty dict
+        is returned
+
+        Returns
+        -------
+        references : dict
+            Dictionary of reference species for the density, temperature
+            and mass along with reference magnetic field and length. The
+            electron temp, density and ratio of R_geometric/R_major is
+            included where R_geometric corresponds to the R where Bref is.
+            B0 means magnetic field at the centre of the local flux surface
+            and Bgeo is the magnetic field at the centre of the last closed
+            flux surface.
+        """
+
+        default_references = {
+            "nref_species": "electron",
+            "tref_species": "electron",
+            "mref_species": "deuterium",
+            "bref": "B0",
+            "lref": "major_radius",
+            "ne": 1.0,
+            "te": 1.0,
+            "vref": "most_probable",
+        }
+
+        references = copy.copy(default_references)
+
+        dens_index = []
+        temp_index = []
+
+        found_electron = False
+        e_mass = None
+        # Get electron temp and density to normalise input
+
+        # Load each species into a dictionary
+        for i_sp in range(self.data["gridsize"]["number_of_species"]):
+            if self.data["species"][i_sp]["z"] == -1:
+                references["ne"] = self.data["species"][i_sp]["dens"]
+                references["te"] = self.data["species"][i_sp]["temp"]
+                electron_index = i_sp
+                e_mass = self.data["species"][i_sp]["mass"]
+                found_electron = True
+
+            # Find all reference values
+            if np.isclose(self.data["species"][i_sp]["dens"], 1.0):
+                dens_index.append(i_sp)
+            if np.isclose(self.data["species"][i_sp]["temp"], 1.0):
+                temp_index.append(i_sp)
+
+        if not found_electron:
+            ne = 0.0
+            for i_sp in range(self.data["gridsize"]["number_of_species"]):
+                ne += (
+                    self.data["species"][i_sp]["dens"] * self.data["species"][i_sp]["z"]
+                )
+
+            references["ne"] = ne
+            references["te"] = self.data["species"][0]["temp"]
+
+        if len(temp_index) == 0 or len(dens_index) == 0:
+            raise ValueError("Cannot find any reference temperature/density species")
+
+        me_md = (electron_mass / deuterium_mass).m
+        me_mh = (electron_mass / hydrogen_mass).m
+
+        if np.isclose(e_mass, 1.0):
+            references["mref_species"] = "electron"
+        elif np.isclose(e_mass, me_md, rtol=0.1):
+            references["mref_species"] = "deuterium"
+        elif np.isclose(e_mass, me_mh, rtol=0.1):
+            references["mref_species"] = "hydrogen"
+        else:
+            raise ValueError("Cannot determine reference mass")
```

Comment:
To-do for future PR: pull out as helper function?

Reply:
Added to match other codes and included in tests